src/System.Private.ServiceModel/src/System/ServiceModel/Channels/RequestChannel.cs

```diff
@@ -185,28 +187,36 @@ protected override void OnAbort()
 
         private void ReleaseRequest(IRequestBase request)
         {
-            if (request != null)
+            try
             {
-                // Synchronization of OnReleaseRequest is the 
-                // responsibility of the concrete implementation of request.
-                request.OnReleaseRequest();
+                if (request != null)
+                {
+                    // Synchronization of OnReleaseRequest is the 
+                    // responsibility of the concrete implementation of request.
+                    request.OnReleaseRequest();
+                }
             }
-
-            lock (_outstandingRequests)
+            finally
             {
-                // Remove supports the connection having been removed, so don't need extra Contains() check,
-                // even though this may have been removed by Abort()
-                _outstandingRequests.Remove(request);
-                if (_outstandingRequests.Count == 0)
+                // Setting _closedTcs needs to happen in a finally block to guarantee that we complete
+                // a waiting close even if OnReleaseRequest throws
+                lock (_outstandingRequests)
                 {
-                    // When we are closed or closing, _closedTcs is managed by the close logic.
-                    if (!_closed && State != CommunicationState.Closing)
+                    _outstandingRequests.Remove(request);
+                    var outstandingRequestCloseCount = Interlocked.Decrement(ref _outstandRequestCloseCount);
+
+                    if (outstandingRequestCloseCount == 0 && _closedTcs != null)
                     {
-                        // Protect against close altering _closedTcs concurrently
-                        var closedTcs = Interlocked.CompareExchange(ref _closedTcs, null, _closedTcs);
-                        if (closedTcs != null)
+                        // When we are closed or closing, _closedTcs is managed by the close logic.
+                        if (!_closed)
```

Comment:
Nit (not required to merge): I think you could combine this 'if (!_closed)' test with the if above.